### PR TITLE
Issue #7744: Update AbstractChecks to log DetailAST - MissingCtor

### PR DESCRIPTION
**Pull request by HuGanghui**
Issue #7744: Update AbstractChecks to log DetailAST - MissingCtor

https://github.com/checkstyle/checkstyle/blob/5ec8cfe4ded197c67f0f4bb8edfddde1d34d78c1/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingCtorCheck.java#L71

Diff report:
https://huganghui.github.io/7744-DetailAST-missing-ctor/reports/diff/

### CODE REVIEW

**Comment by HuGanghui** [comment]
@rnveach I have uploaded diff report.

**Comment by rnveach** [comment]
https://huganghui.github.io/7744-DetailAST-missing-ctor/reports/diff/checkstyle/index.html
> Number of unique base messages reported below: 2836
Number of unique patch messages reported below: 2838

Counts don't match up as was mentioned in the parent issue for this.

Please find the new violations that weren't being reported before and ensure we have them as directly part of our MissingCtor tests.

**Comment by romani** [comment]
all other reports looks ok.

+2  cases in checkstyle might be due to introduction of two new Tests Inputs for Xpath. But for some reason I can not find them in report page

**Comment by romani** [comment]
@rnveach , we might need to update our diff site generation to publish property files also, to troubleshoot contributors problems easily .

**Comment by HuGanghui** [comment]
@romani  I find them in diff report:

image

image

image

image

**Comment by HuGanghui** [comment]
And they appear again when I try to know why the new violations happen:

make the three class in different lines:

cat Test.java

```java
package checkstyle_demo.ast_missingctor;

public class Test {
    static class Inner1 {
        static class Inner2<V> {
            public void m() {

            }
        }
    }
}
```
cat Checkstyle_cmd.sh
```shell 
#!/usr/bin/env bash
 RUN_LOCALE="-Duser.language=en -Duser.country=US"
// origin jar 
 java $RUN_LOCALE -jar /var/tmp/checkstyle-8.31-SNAPSHOT-all.jar -c config.xml Test.java
// update to log jar
 java $RUN_LOCALE -jar /Users/hgh/Desktop/checkstyle/target/checkstyle-8.31-SNAPSHOT-all.jar -c config.xml Test.java
```

Starting audit...
[ERROR] /Users/hgh/Downloads/my_file/IDEA_workspace/Leetcode_maven/engineering_ability/src/main/java/checkstyle_demo/ast_missingctor/Test.java:3: Class should define a constructor. [MissingCtor]
[ERROR] /Users/hgh/Downloads/my_file/IDEA_workspace/Leetcode_maven/engineering_ability/src/main/java/checkstyle_demo/ast_missingctor/Test.java:4: Class should define a constructor. [MissingCtor]
[ERROR] /Users/hgh/Downloads/my_file/IDEA_workspace/Leetcode_maven/engineering_ability/src/main/java/checkstyle_demo/ast_missingctor/Test.java:5: Class should define a constructor. [MissingCtor]
Audit done.
Checkstyle ends with 3 errors.
Starting audit...
[ERROR] /Users/hgh/Downloads/my_file/IDEA_workspace/Leetcode_maven/engineering_ability/src/main/java/checkstyle_demo/ast_missingctor/Test.java:3:1: Class should define a constructor. [MissingCtor]
[ERROR] /Users/hgh/Downloads/my_file/IDEA_workspace/Leetcode_maven/engineering_ability/src/main/java/checkstyle_demo/ast_missingctor/Test.java:4:5: Class should define a constructor. [MissingCtor]
[ERROR] /Users/hgh/Downloads/my_file/IDEA_workspace/Leetcode_maven/engineering_ability/src/main/java/checkstyle_demo/ast_missingctor/Test.java:5:9: Class should define a constructor. [MissingCtor]
Audit done.
Checkstyle ends with 3 errors.

When I make the three class in the same line, something different happen: 

cat Test.java

```java
package checkstyle_demo.ast_missingctor;

public class Test { static class Inner1 { static class Inner2<V> {
            public void m() {

            }
        }
    }
}
```
cat Checkstyle_cmd.sh

```shell 
#!/usr/bin/env bash
 RUN_LOCALE="-Duser.language=en -Duser.country=US"
// origin jar 
 java $RUN_LOCALE -jar /var/tmp/checkstyle-8.31-SNAPSHOT-all.jar -c config.xml Test.java
// update to log jar
 java $RUN_LOCALE -jar /Users/hgh/Desktop/checkstyle/target/checkstyle-8.31-SNAPSHOT-all.jar -c config.xml Test.java
```

Starting audit...
[ERROR] /Users/hgh/Downloads/my_file/IDEA_workspace/Leetcode_maven/engineering_ability/src/main/java/checkstyle_demo/ast_missingctor/Test.java:3: Class should define a constructor. [MissingCtor]
Audit done.
Checkstyle ends with 1 errors.
Starting audit...
[ERROR] /Users/hgh/Downloads/my_file/IDEA_workspace/Leetcode_maven/engineering_ability/src/main/java/checkstyle_demo/ast_missingctor/Test.java:3:1: Class should define a constructor. [MissingCtor]
[ERROR] /Users/hgh/Downloads/my_file/IDEA_workspace/Leetcode_maven/engineering_ability/src/main/java/checkstyle_demo/ast_missingctor/Test.java:3:21: Class should define a constructor. [MissingCtor]
[ERROR] /Users/hgh/Downloads/my_file/IDEA_workspace/Leetcode_maven/engineering_ability/src/main/java/checkstyle_demo/ast_missingctor/Test.java:3:43: Class should define a constructor. [MissingCtor]
Audit done.
Checkstyle ends with 3 errors.

I think maybe the reason is that violation message of the same check in one line such as MissingCtor is the same, and it will be ignored in origin jar where violation message doesn't contain ColumnNo. So this problem did't happen when code update to log and have ColumnNo.

**Comment by rnveach** [comment]
> I think maybe the reason is that violation message of the same check in one line such as MissingCtor is the same, and it will be ignored in origin jar where violation message doesn't contain ColumnNo. So this problem did't happen when code update to log and have ColumnNo.

**Comment by rnveach** [comment]
> I think maybe the reason is that violation message of the same check in one line such as MissingCtor is the same, and it will be ignored in origin jar where violation message doesn't contain ColumnNo. So this problem did't happen when code update to log and have ColumnNo.

Correct. So please ensure we have them as directly part of our MissingCtor tests. If we already have examples then just point me to them and I can approve this PR.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingCtorCheckTest.java [hunk 1]
```diff
@@ -56,4 +56,20 @@ public void testTokensNotNull() {
...
+            "3:1: " + getCheckMessage(MSG_KEY),
+            "3:34: " + getCheckMessage(MSG_KEY),
+            "3:49: " + getCheckMessage(MSG_KEY),
+        };
+
+        verify(checkConfig,
+                getPath("InputMissingCtor2.java"),
+                expected);
+    }
+
```
Comment (HuGanghui) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingCtorCheckTest.java]: @rnveach I add a UT for this situation.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/missingctor/InputMissingCtor2.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package com.puppycrawl.tools.checkstyle.checks.coding.missingctor;
+
+public class InputMissingCtor2 { class Inner1 { class Inner2 {
+
+        }
+    }
+}
```
Comment (HuGanghui) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/missingctor/InputMissingCtor2.java]: @rnveach the example added.

**Pull request merged**